packages/vite/src/node/ssr/runtime/runtime.ts

```diff
@@ -0,0 +1,329 @@
+import type { ViteHotContext } from 'types/hot'
+import { HMRClient, HMRContext } from '../../../shared/hmr'
+import { ModuleCacheMap } from './moduleCache'
+import type {
+  FetchResult,
+  ImportMetaEnv,
+  ModuleCache,
+  ResolvedResult,
+  SSRImportMetadata,
+  ViteModuleRunner,
+  ViteRuntimeImportMeta,
+  ViteRuntimeModuleContext,
+  ViteServerClientOptions,
+} from './types'
+import {
+  cleanUrl,
+  createImportMetaEnvProxy,
+  isPrimitive,
+  isWindows,
+  posixDirname,
+  posixPathToFileHref,
+  posixResolve,
+  toWindowsPath,
+  unwrapId,
+} from './utils'
+import {
+  ssrDynamicImportKey,
+  ssrExportAllKey,
+  ssrImportKey,
+  ssrImportMetaKey,
+  ssrModuleExportsKey,
+} from './constants'
+import { ViteRuntimeHMRLogger, ViteRuntimeSilentHMRLogger } from './hmrLogger'
+
+interface ViteRuntimeDebugger {
+  (formatter: unknown, ...args: unknown[]): void
+}
+
+export class ViteRuntime {
+  /**
+   * Holds the cache of modules
+   * Keys of the map are ids
+   */
+  public moduleCache: ModuleCacheMap
+  public hmrClient?: HMRClient
+  public entrypoints = new Set<string>()
+
+  private idToFileMap = new Map<string, string>()
+  private envProxy: ImportMetaEnv
+
+  constructor(
+    public options: ViteServerClientOptions,
+    public runner: ViteModuleRunner,
+    private debug?: ViteRuntimeDebugger,
+  ) {
+    this.moduleCache = options.moduleCache ?? new ModuleCacheMap(options.root)
+    this.envProxy = createImportMetaEnvProxy(options.environmentVariables)
+    if (typeof options.hmr === 'object') {
+      this.hmrClient = new HMRClient(
+        options.hmr.logger === false
+          ? new ViteRuntimeSilentHMRLogger()
+          : options.hmr.logger || new ViteRuntimeHMRLogger(),
+        options.hmr.connection,
+        ({ acceptedPath, ssrInvalidates }) => {
+          this.moduleCache.delete(acceptedPath)
+          ssrInvalidates?.forEach((id) => this.moduleCache.delete(id))
+          return this.executeUrl(acceptedPath)
+        },
+      )
+    }
+  }
+
+  /**
+   * URL to execute. Accepts file path, server path or id relative to the root.
+   */
+  public async executeUrl<T = any>(url: string): Promise<T> {
+    const fetchedModule = await this.cachedModule(url)
+    return await this.cachedRequest(url, fetchedModule, [])
+  }
+
+  /**
+   * Entrypoint URL to execute. Accepts file path, server path or id relative to the root.
+   * In the case of a full reload triggered by HMR, these are the modules that will be reloaded
+   */
+  public async executeEntrypoint<T = any>(url: string): Promise<T> {
```

Comment:
There are two public methods. The only difference is that `executeEntrypoint` supports a full-reload event (it's the file that will be reloaded). Can we assume that all public methods should behave like that? (So, any path in `executeUrl` call will be reloaded when `full-reload` is triggered) If so, we need only a single API for this